### ISSUE TITLE
master branch fails tests on linux x86_64

### ISSUE THREAD

**Issue by naquad** [ned14/nedmalloc#3]
Hi.

Box:
Distro: Arch Linux
Uname: Linux nq.od.ua 2.6.36-ARCH #1 SMP PREEMPT Sat Jan 8 14:15:27 CET 2011 x86_64 Intel(R) Core(TM)2 Quad CPU Q9550 @ 2.83GHz GenuineIntel GNU/Linux
gLibc: glibc 2.12.2-2
GCC: 4.5.2
Binutils: 2.21-2

Issue:

[000][15:29:52][naquad@nq:~/s0ft/tmp]$ git clone https://github.com/ned14/nedmalloc.git
Cloning into nedmalloc...
remote: Counting objects: 1087, done.
remote: Compressing objects: 100% (366/366), done.
remote: Total 1087 (delta 706), reused 1057 (delta 687)
Receiving objects: 100% (1087/1087), 1.73 MiB | 763 KiB/s, done.
Resolving deltas: 100% (706/706), done.
[000][15:29:59][naquad@nq:~/s0ft/tmp]$ cd nedmalloc/
/home/naquad/s0ft/tmp/nedmalloc
[000][15:30:02][naquad@nq:~/s0ft/tmp/nedmalloc]$ git submodule init
Submodule 'nedtries' (git://github.com/ned14/nedtries.git) registered for path 'nedtries'
[000][15:30:05][naquad@nq:~/s0ft/tmp/nedmalloc]$ git submodule update
Cloning into nedtries...
remote: Counting objects: 151, done.
remote: Compressing objects: 100% (75/75), done.
remote: Total 151 (delta 75), reused 151 (delta 75)
Receiving objects: 100% (151/151), 21.22 MiB | 1.45 MiB/s, done.
Resolving deltas: 100% (75/75), done.
Submodule path 'nedtries': checked out 'f45eadddc36f32f4742980a0fc6a0627a2789e20'
[000][15:30:24][naquad@nq:~/s0ft/tmp/nedmalloc]$ scons
scons: Reading SConscript files ...
**\* Build variant preferred by environment is Release x64
scons: done reading SConscript files.
scons: Building targets ...
g++ -o x64/Release/test_cpp.o -c -std=c++0x -fstrict-aliasing -fargument-noalias -Wstrict-aliasing -Wall -Wno-unused -O2 -g -DENABLE_TOLERANT_NEDMALLOC -DUSE_ALLOCATOR=1 -DUSE_LOCKS=1 -DNDEBUG test.cpp
In file included from test.cpp:20:0:
nedmalloc.c: In function 'malloc_state\* nedalloc::FindMSpace(nedalloc::nedpool_, nedalloc::threadcache_, int_, size_t)':
nedmalloc.c:1582:65: warning: operation on 'p->nedalloc::nedpool_t::m[end]' may be undefined
gcc -o x64/Release/nedmalloc.os -c -fstrict-aliasing -fargument-noalias -Wstrict-aliasing -Wall -Wno-unused -O2 -g -fPIC -DENABLE_TOLERANT_NEDMALLOC -DUSE_ALLOCATOR=1 -DUSE_LOCKS=1 -DNDEBUG -DNEDMALLOC_DLL_EXPORTS nedmalloc.c
gcc -o x64/Release/libnedmalloc_test.so -shared x64/Release/nedmalloc.os -lrt -lpthread
g++ -o x64/Release/test_cpp x64/Release/test_cpp.o -lrt -lpthread x64/Release/libnedmalloc_test.so
gcc -o x64/Release/test_c.o -c -fstrict-aliasing -fargument-noalias -Wstrict-aliasing -Wall -Wno-unused -O2 -g -DENABLE_TOLERANT_NEDMALLOC -DUSE_ALLOCATOR=1 -DUSE_LOCKS=1 -DNDEBUG test.c
gcc -o x64/Release/test_c x64/Release/test_c.o -lrt -lpthread x64/Release/libnedmalloc_test.so
g++ -o x64/Release/scalingtest.o -c -std=c++0x -fstrict-aliasing -fargument-noalias -Wstrict-aliasing -Wall -Wno-unused -O2 -g -DENABLE_TOLERANT_NEDMALLOC -DUSE_ALLOCATOR=1 -DUSE_LOCKS=1 -DNDEBUG scalingtest.cpp
In file included from scalingtest.cpp:13:0:
nedmalloc.c: In function 'malloc_state_ nedalloc::FindMSpace(nedalloc::nedpool_, nedalloc::threadcache_, int*, size_t)':
nedmalloc.c:1582:65: warning: operation on 'p->nedalloc::nedpool_t::m[end]' may be undefined
g++ -o x64/Release/scalingtest x64/Release/scalingtest.o -lrt -lpthread
gcc -o x64/Release/make_pgos.o -c -fstrict-aliasing -fargument-noalias -Wstrict-aliasing -Wall -Wno-unused -O2 -g -DENABLE_TOLERANT_NEDMALLOC -DUSE_ALLOCATOR=1 -DUSE_LOCKS=1 -DNDEBUG make_pgos.c
gcc -o x64/Release/make_pgos x64/Release/make_pgos.o -lrt -lpthread x64/Release/libnedmalloc_test.so
scons: done building targets.
[000][15:30:35][naquad@nq:~/s0ft/tmp/nedmalloc]$ LANG=C ./x64/Release/test_c
Running torture test
-=-=-=-=-=-=-=-=-=-=
Block size <= 2097152, C++ test mode is on

Testing nedmalloc with 1 threads ...
Bus error

gdb invocation:

[000][15:32:10][naquad@nq:~/s0ft/tmp/nedmalloc]$ make debug
scons -k --debugbuild
scons: Reading SConscript files ...
**\* Build variant preferred by environment is Debug x64
scons: done reading SConscript files.
scons: Building targets ...
g++ -o x64/Debug/test_cpp.o -c -std=c++0x -fstrict-aliasing -fargument-noalias -Wstrict-aliasing -Wall -Wno-unused -O0 -g -DENABLE_TOLERANT_NEDMALLOC -DUSE_ALLOCATOR=1 -DUSE_LOCKS=1 -DDEBUG -D_DEBUG test.cpp
In file included from test.cpp:20:0:
nedmalloc.c: In function 'malloc_state\* nedalloc::FindMSpace(nedalloc::nedpool_, nedalloc::threadcache_, int_, size_t)':
nedmalloc.c:1582:65: warning: operation on 'p->nedalloc::nedpool_t::m[end]' may be undefined
gcc -o x64/Debug/nedmalloc.os -c -fstrict-aliasing -fargument-noalias -Wstrict-aliasing -Wall -Wno-unused -O0 -g -fPIC -DENABLE_TOLERANT_NEDMALLOC -DUSE_ALLOCATOR=1 -DUSE_LOCKS=1 -DDEBUG -D_DEBUG -DNEDMALLOC_DLL_EXPORTS nedmalloc.c
gcc -o x64/Debug/libnedmalloc_test.so -shared x64/Debug/nedmalloc.os -lrt -lpthread
g++ -o x64/Debug/test_cpp x64/Debug/test_cpp.o -lrt -lpthread x64/Debug/libnedmalloc_test.so
gcc -o x64/Debug/test_c.o -c -fstrict-aliasing -fargument-noalias -Wstrict-aliasing -Wall -Wno-unused -O0 -g -DENABLE_TOLERANT_NEDMALLOC -DUSE_ALLOCATOR=1 -DUSE_LOCKS=1 -DDEBUG -D_DEBUG test.c
gcc -o x64/Debug/test_c x64/Debug/test_c.o -lrt -lpthread x64/Debug/libnedmalloc_test.so
g++ -o x64/Debug/scalingtest.o -c -std=c++0x -fstrict-aliasing -fargument-noalias -Wstrict-aliasing -Wall -Wno-unused -O0 -g -DENABLE_TOLERANT_NEDMALLOC -DUSE_ALLOCATOR=1 -DUSE_LOCKS=1 -DDEBUG -D_DEBUG scalingtest.cpp
In file included from scalingtest.cpp:13:0:
nedmalloc.c: In function 'malloc_state_ nedalloc::FindMSpace(nedalloc::nedpool_, nedalloc::threadcache_, int*, size_t)':
nedmalloc.c:1582:65: warning: operation on 'p->nedalloc::nedpool_t::m[end]' may be undefined
g++ -o x64/Debug/scalingtest x64/Debug/scalingtest.o -lrt -lpthread
gcc -o x64/Debug/make_pgos.o -c -fstrict-aliasing -fargument-noalias -Wstrict-aliasing -Wall -Wno-unused -O0 -g -DENABLE_TOLERANT_NEDMALLOC -DUSE_ALLOCATOR=1 -DUSE_LOCKS=1 -DDEBUG -D_DEBUG make_pgos.c
gcc -o x64/Debug/make_pgos x64/Debug/make_pgos.o -lrt -lpthread x64/Debug/libnedmalloc_test.so
scons: done building targets.
[000][15:32:15][naquad@nq:~/s0ft/tmp/nedmalloc]$ ./x64/Debug/test_c
Running torture test
-=-=-=-=-=-=-=-=-=-=
Block size <= 2097152, C++ test mode is on

Testing nedmalloc with 1 threads ...
Bus error
[135][15:32:24][naquad@nq:~/s0ft/tmp/nedmalloc]$ gdb ./x64/Debug/test_c
GNU gdb (GDB) 7.2
Copyright (C) 2010 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later http://gnu.org/licenses/gpl.html
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-unknown-linux-gnu".
For bug reporting instructions, please see:
http://www.gnu.org/software/gdb/bugs/...
Reading symbols from /home/naquad/s0ft/tmp/nedmalloc/x64/Debug/test_c...done.
(gdb) r
Starting program: /home/naquad/s0ft/tmp/nedmalloc/x64/Debug/test_c 
[Thread debugging using libthread_db enabled]
Running torture test
-=-=-=-=-=-=-=-=-=-=
Block size <= 2097152, C++ test mode is on

Testing nedmalloc with 1 threads ...
[New Thread 0x7ffff7444700 (LWP 7885)]

Program received signal SIGBUS, Bus error.
[Switching to Thread 0x7ffff7444700 (LWP 7885)]
0x00007ffff77a4eb4 in mmap_alloc (m=0x7ffff7ed0010, nb=641712, flags=0) at malloc.c.h:4091
4091          _(void_*)mm = mmaph;
(gdb) bt
#0  0x00007ffff77a4eb4 in mmap_alloc (m=0x7ffff7ed0010, nb=641712, flags=0) at malloc.c.h:4091
#1  0x00007ffff77a676c in sys_alloc (m=0x7ffff7ed0010, nb=641712, flags=0) at malloc.c.h:4299
#2  0x00007ffff77b0027 in mspace_malloc_implementation (msp=0x7ffff7ed0010, bytes=641692, alignment=0, flags=0) at malloc.c.h:5574
#3  mspace_malloc2 (msp=0x7ffff7ed0010, bytes=641692, alignment=0, flags=0) at malloc.c.h:5599
#4  0x00007ffff77b5d04 in CallMalloc (p=0x7ffff79b9560, size=641692, alignment=0, flags=0) at nedmalloc.c:331
#5  nedpmalloc2 (p=0x7ffff79b9560, size=641692, alignment=0, flags=0) at nedmalloc.c:1899
#6  0x00007ffff77b6db2 in nedpmalloc (p=0x0, size=641692) at nedmalloc.c:2008
#7  0x00007ffff77b2c47 in nedmalloc (size=641692) at nedmalloc.c:626
#8  0x0000000000400ef8 in threadcode (threadidx=0) at test.c:216
#9  0x0000000000400cfb in _threadcode (a=0x0) at test.c:147
#10 0x00007ffff79c0cb0 in start_thread () from /lib/libpthread.so.0
#11 0x00007ffff75137ed in clone () from /lib/libc.so.6
#12 0x0000000000000000 in ?? ()

(gdb)

**Comment by naquad** [comment]
sorry for formatting, tried to use pre and code tags, but half of data is lost with them :(

**Comment by ned14** [comment]
Firstly it's easier to attach printouts using the instructions in the sticky http://github.com/ned14/nedmalloc/issues#issue/1.

So apparently it's bus faulting when trying to mmap some new memory on Linux on x64? I am very surprised. I only had that running very recently here without issue.

I'll get back to you when I get a chance to run it here.

Thanks for reporting this issue.

**Comment by ned14** [comment]
Hi,

My deepest apologies for such a long wait. Copying exactly what you posted, I am afraid to say it works absolutely fine here. Here's my system config:

Kubtunu 10.10 with kernel 2.6.35-24-generic
GCC v4.5.1
binutils 2.20.51

Here's my output:

```
ned@kate-linux:~/Tornado/Tn/TClient/TnFOX/src/nedmalloc$ scons
scons: Reading SConscript files ...
*** Build variant preferred by environment is Release x64
scons: done reading SConscript files.
scons: Building targets ...
g++ -o x64/Release/test_cpp.o -c -std=c++0x -fstrict-aliasing -fargument-noalias -Wstrict-aliasing -Wall -Wno-unused -O2 -g -DENABLE_TOLERANT_NEDMALLOC -DUSE_ALLOCATOR=1 -DUSE_LOCKS=1 -DNDEBUG test.cpp
gcc -o x64/Release/nedmalloc.os -c -fstrict-aliasing -fargument-noalias -Wstrict-aliasing -Wall -Wno-unused -O2 -g -fPIC -DENABLE_TOLERANT_NEDMALLOC -DUSE_ALLOCATOR=1 -DUSE_LOCKS=1 -DNDEBUG -DNEDMALLOC_DLL_EXPORTS nedmalloc.c
gcc -o x64/Release/libnedmalloc_test.so -shared x64/Release/nedmalloc.os -lrt -lpthread
g++ -o x64/Release/test_cpp x64/Release/test_cpp.o -lrt -lpthread x64/Release/libnedmalloc_test.so
gcc -o x64/Release/test_c.o -c -fstrict-aliasing -fargument-noalias -Wstrict-aliasing -Wall -Wno-unused -O2 -g -DENABLE_TOLERANT_NEDMALLOC -DUSE_ALLOCATOR=1 -DUSE_LOCKS=1 -DNDEBUG test.c
gcc -o x64/Release/test_c x64/Release/test_c.o -lrt -lpthread x64/Release/libnedmalloc_test.so
g++ -o x64/Release/scalingtest.o -c -std=c++0x -fstrict-aliasing -fargument-noalias -Wstrict-aliasing -Wall -Wno-unused -O2 -g -DENABLE_TOLERANT_NEDMALLOC -DUSE_ALLOCATOR=1 -DUSE_LOCKS=1 -DNDEBUG scalingtest.cpp
g++ -o x64/Release/scalingtest x64/Release/scalingtest.o -lrt -lpthread
gcc -o x64/Release/make_pgos x64/Release/make_pgos.o -lrt -lpthread x64/Release/libnedmalloc_test.so
scons: done building targets.
ned@kate-linux:~/Tornado/Tn/TClient/TnFOX/src/nedmalloc$ x64/Release/test_c
Running torture test
-=-=-=-=-=-=-=-=-=-=
Block size <= 2097152, C++ test mode is on

Testing nedmalloc with 1 threads ...
Relaunched thread 0
Relaunched thread 0
Relaunched thread 0
Relaunched thread 0
Relaunched thread 0
Relaunched thread 0
Relaunched thread 0
Relaunched thread 0
This test spent 6.094797% of its time doing reallocs
This allocator achieves 791.170623ops/sec under 1 threads

Press a key to trim

ned@kate-linux:~/Tornado/Tn/TClient/TnFOX/src/nedmalloc$ 
```

Note that the very low score is because I am running VirtualBox as I currently don't have a native Linux install available to me. Surely that wouldn't make a difference though?

Niall

**Comment by naquad** [comment]
I've retried again (git clone, git submodule update, scons ...), but still
got this issue.
Anyway, I've implemented my own memory manager for my program based on your
approach.
Thank you for this project, some things were unclear to me but I've figured
out everything I've needed.

2011/5/19 ned14 <
reply@reply.github.com>

**Comment by ned14** [comment]
There was an issue last year where due to a bad granularity aligned implementation on POSIX it would randomly fail to map new memory and that would trigger the error you saw in debug builds only. However that was fixed in 4cca52c4b8b934698048, and given that that was in December and you had the problem in release only I thought it couldn't be that.

I am sorry that you had to come up with your own implementation. Kinda surprised you didn't reach for ptmalloc3 as if you're on Linux only it's ready to go out of the box. And it's swift.

I'm releasing v1.10 beta1 today. Beta 2 is going to be very different, a whole new dlmalloc, new API based on the latest movements in ISO C1X and more.

**Comment by ned14** [comment]
I do apologise - there was indeed a bug here fixed in issue #10. Your segfault was caused by a failure to correctly trap a failure to request memory from the system.
